docs: refactor live-patching database docs. (#558)
canonical/launchpad-manual: docs/developer/explanation/database/live-patching.rst

that, before landing a database patch, we may need to land code first so that
everything runs both with and without the patch applied. For the end-to-end
workflow, see :ref:`database-schema-changes-process`.

We land cold schema changes on ``db-devel`` and hot ones on ``devel``. All
schema changes must land without appserver code changes, although test changes
may be included when trigger or database function changes are being landed. For
more on developing against the ``db-devel`` branch, see
:ref:`working-with-db-devel`.

.. note::

   These instructions only apply to tables under native PostgreSQL streaming
   replication. ``lp_*`` tables are replicated to SSO via Slony, so they
   require extra special care. See :ref:`postgresql-and-launchpad` for more on
   how Launchpad uses replication.

Hot patches
-----------

We can apply some database patches live. We currently apply them by crafting an
appropriate wrapper script and then running it directly on the master database
node, from where it replicates to the slaves.

The ``(major, minor, patch)`` revision tuple must be manually inserted into the
``LaunchpadDatabaseRevision`` table on the master. Failing to do this aborts the

> next full upgrade, because the system attempts to apply a patch that has already

Reviewer: ```suggestion
next full upgrade because the system attempts to apply a patch that has already
```

been applied. This situation is normally caught by the full staging update.

We intend to automate this in the future, to at least handle the happy path and
avoid accidentally running a hot patch during a fastdowntime deployment.

Index creation
~~~~~~~~~~~~~~

Index creation for all but the smallest tables should be done in a hot patch.

The patch must use ``CREATE INDEX CONCURRENTLY``, which runs three separate fast
transactions to create, populate, and activate the index. This adds some
contention, but it is minimal.

Don't start index creation while there are long-running transactions in

> progress: the index will not complete until those transactions have completed,

Reviewer: ```suggestion
progress: the index will not complete until those transactions have completed
```

because it has to wait for the existing read transactions on the table to finish
before it can be activated.

``CREATE INDEX CONCURRENTLY`` cannot run inside a transaction, so if the index

> cannot be built for some reason, such as trying to build a ``UNIQUE`` index on

Reviewer: ```suggestion
``CREATE INDEX CONCURRENTLY`` cannot run inside a transaction. If the index
cannot be built for some reason, such as trying to build a ``UNIQUE`` index on
```

> non-unique data, manual repair is needed to back out the changes.

Reviewer: Does `back out` here mean roll back? If so, we should keep the terminology consistent across the docs.

Author: yes

Author: ```suggestion
non-unique data, manual repair is needed to roll back the changes.
```

.. warning::

>    Deleting an index requires a fastdowntime deploy.

Reviewer: ```suggestion
   Deleting an index requires a fastdowntime deployment.
```

Run index creation from a ``screen(1)`` session or similar to avoid network

> dropouts: for ``CREATE INDEX CONCURRENTLY``, an early disconnection results in

Reviewer: ```suggestion
dropouts. For ``CREATE INDEX CONCURRENTLY``, an early disconnection results in
```

an invalid index rather than a rollback.

Once the index is built on the master, it replicates automatically to the
slaves. After creating the indexes, confirm that there are no invalid indexes 
on the affected tables.

Views
~~~~~

Run the ``CREATE OR REPLACE VIEW`` statements on the master.

Stored procedures
~~~~~~~~~~~~~~~~~

Stored procedures can be changed, as long as the signature is still consistent
with any triggers. Table trigger definitions must be done as cold patches.

Cold patches
------------

> Cold patches must be applied with no other system activity, to prevent locks

Reviewer: ```suggestion
Cold patches must be applied with no other system activity to prevent locks
```

an invalid index rather than a rollback.

After creating the indexes, confirm that there are no invalid indexes on
the affected tables.
Once the index is built on the master, it replicates automatically to the
slaves. After creating the indexes, confirm that there are no invalid indexes 
on the affected tables.

Views
~~~~~

Run the CREATE OR REPLACE VIEW statements on the master.
Run the ``CREATE OR REPLACE VIEW`` statements on the master.

Stored procedures
~~~~~~~~~~~~~~~~~

Stored procedures can be changed (as long as the signature is still
consistent with any triggers).

Table trigger definitions must be done as cold patches.
Stored procedures can be changed, as long as the signature is still consistent
with any triggers. Table trigger definitions must be done as cold patches.

Cold patches
------------

Cold patches need to be applied with no other system activity to prevent
locks holding the patch up. We have automation around this to do it as
Cold patches must be applied with no other system activity, to prevent locks
from holding the patch up. We have automation around this to apply them as
rapidly as possible.

Patch process overview
~~~~~~~~~~~~~~~~~~~~~~

> 1. Update wildcherry Launchpad source tree to the revision to run: make

Author: I am not sure if this bit is still relevant.

Reviewer 2: I'm not aware of `wildcherry`

While it's true that we apply one patch at a time, there's a doc that explains how to run multiple database schema changes (in different downtime windows).

>    at a time to reduce the debugging burden when something goes wrong after

Reviewer: ```suggestion
   at a time to reduce the debugging burden if something goes wrong after
```
A little optimism 😄

> 3. Run ``full-update.py``.

Reviewer 2: we don't use this script now, we use the juju actions: `run-action launchpad-db-update db-update`

Author: I think this refers to applying the patch locally. `make clean build` does not make sense for non-prod and prod envs.

I'm +1 to drop this section entirely as we are covering the actual process to setup a patch in other parts.

> key relations need exclusive table locks to activate. Do it as a cold patch.

Reviewer: ```suggestion
Creating a table can sometimes be done hot, but not reliably. It should be done as a cold patch because foreign key relations need exclusive table locks to activate.
```

>    constraints to a busy table, which requires locks on the busy table that

Reviewer: ```suggestion
   constraints to a busy table which requires locks on the busy table that
```
